rs/rosetta-api/icrc1/ledger/sm-tests/src/lib.rs

```diff
@@ -2464,9 +2469,129 @@ pub fn icrc1_test_block_transformation<T>(
+                            &transactions[tx_index],
+                            TimeStamp::from_nanos_since_unix_epoch(system_time_to_nanos(
+                                env.time(),
+                            )),
+                            minter_principal,
+                            Some(FEE.into()),
+                        );
+                        tx_index += 1;
+                    }
+                    tx_index_target += ADDITIONAL_TX_BATCH_SIZE;
+                    in_memory_ledger.verify_balances_and_allowances(&env, ledger_id);
+                };
+                add_tx_and_verify();
+
+                let mut test_upgrade = |ledger_wasm: Vec<u8>| {
+                    env.upgrade_canister(ledger_id, ledger_wasm, upgrade_args.clone())
+                        .unwrap();
+                    add_tx_and_verify();
+                };
+
+                // Test if the old serialized approvals and balances are correctly deserialized
+                test_upgrade(ledger_wasm_current.clone());
+                // Test the new wasm serialization
+                test_upgrade(ledger_wasm_current.clone());
+                if let Some(ledger_wasm_upgradetomemorymanager) =
+                    ledger_wasm_upgradetomemorymanager.clone()
+                {
+                    // Test serializing to the memory manager
+                    test_upgrade(ledger_wasm_upgradetomemorymanager.clone());
+                    // Test upgrade to memory manager again
+                    test_upgrade(ledger_wasm_upgradetomemorymanager);
+
+                    // Current mainnet wasm cannot deserialize from memory manager
+                    match env.upgrade_canister(
+                        ledger_id,
+                        ledger_wasm_mainnet.clone(),
+                        upgrade_args.clone(),
+                    ) {
+                        Ok(_) => {
+                            panic!("Upgrade from memory manager directly to mainnet should fail!")
+                        }
+                        Err(e) => {
+                            assert!(e.description().contains("failed to decode ledger state"))
+                        }
+                    };
+                }
+                // Test deserializing from memory manager
+                test_upgrade(ledger_wasm_current.clone());
+                // Test downgrade to mainnet wasm
+                test_upgrade(ledger_wasm_mainnet.clone());
+                if verify_blocks {
+                    // This will also verify the ledger blocks.
+                    // The current implementation of the InMemoryLedger cannot get blocks
+                    // for the ICP ledger. This part of the test runs only for the ICRC1 ledger.
+                    verify_ledger_state(&env, ledger_id);
+                }
+
+                Ok(())
+            },
+        )
+        .unwrap();
+}
+
+pub fn icrc1_test_upgrade_serialization_deterministic<T>(
```

Comment:
What does "deterministic" mean here? You don't test if the raw stable memory bytes are always the same.

Reply:
I was referring to the manual generation of transfers and approvals as opposed to the randomly generated transactions in the proptest above. I am not sure if it is a good name though...

Reply:
maybe `icrc1_test_upgrade_serialization_manual` or `icrc1_test_upgrade_serialization_fixed_tx` ?